File: Yandex/Practicum/Algorightms_and_Data_Structures/1_sprint/C.Соседи/solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>

void solution(int n, int m, std::vector<std::vector<int>>& matrix, int row, int col) {
    /*
    Time Complexity O(1)
    Memory Complexity O(1)
    */
    std::vector<int> neighbors;
    std::vector<std::pair<int, int>> dirs = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    for (std::pair<int, int>& dir : dirs) {
        int nextRow = row + dir.first;
        int nextCol = col + dir.second;
        if (0 <= std::min(nextRow, nextCol) && nextRow < n && nextCol < m) {
            neighbors.push_back(matrix[nextRow][nextCol]);
        }
    }
    std::sort(neighbors.begin(), neighbors.end());
    for (int index = 0; index < neighbors.size(); ++index) {
        if (index) {
            std::cout << " ";
        }
        std::cout << neighbors[index];
    }
};


int main() {
    int n, m;
    std::cin >> n >> m;
    std::vector<std::vector<int>> matrix;
    for (int i = 0; i < n; ++i) {
        std::vector<int> row;
        for (int j = 0; j < m; ++j) {
            int num;
            std::cin >> num;
            row.push_back(num);
        }
        matrix.push_back(row);
    }
    int row, col;
    std::cin >> row >> col;
    solution(n, m, matrix, row, col);
}

File: Yandex/Practicum/Algorightms_and_Data_Structures/1_sprint/C.Соседи/solution.py
import sys


def solution(n: int, m: int, matrix: list[list[int]], row: int, col: int) -> str:
    """
    Time Complexity O(1)
    Memory Complexity O(1)
    """
    neighbors: list[int] = []
    dirs: list[tuple[int, int]] = [(0, -1), (0, 1), (-1, 0), (1, 0)]
    for dir_row, dir_col in dirs:
        next_row, next_col = row + dir_row, col + dir_col
        if 0 <= next_row < n and 0 <= next_col < m:
            neighbors.append(matrix[next_row][next_col])
    neighbors.sort()
    return ' '.join(str(neighbor) for neighbor in neighbors)


if __name__ == '__main__':
    n: int = int(sys.stdin.readline().rstrip())
    m: int = int(sys.stdin.readline().rstrip())
    matrix: list[list[int]] = []
    for _ in range(n):
        row: list[int] = list(map(int, sys.stdin.readline().rstrip().split()))
        matrix.append(row)
    row: int = int(sys.stdin.readline().rstrip())
    col: int = int(sys.stdin.readline().rstrip())
    sys.stdout.write(solution(n, m, matrix, row, col))
